### ISSUE TITLE
Convolutional Neural Network for Atari Game: Data Pipeline and Prediction issue

### ISSUE THREAD

**Issue by Raighar** [eclipse/deeplearning4j#4979]
#### Issue Description

I am trying to use a CNN to predict a move policy for an arcade video game. The framework for the game is written in Java and therefore I can directly save the input of the network (board state, character position, etc.) in arrays and provide these to my CNN.
However, I am not 100% sure if I am handling the data correctly because most of the examples only use FileReaders instead of using already existing Java-Arrays. Most of the time, the network only predicts one out of the possible 4 outcomes (UP, DOWN, LEFT, RIGHT) for every test instance.
Before I was using DL4J I had the exact same network built in Keras and I was saving the data to my filesystem and importing it into Python. Surprisingly, after only 10 epochs of training with about 5000 training samples, the network was reaching an accuracy of ~89%.
My DL4J version is running for two days now (Schedule: generating data -> training 10 epochs -> start over). It has by now run through the 10 epochs 70 times but still, the accuracy stays below 55%. I think I might be doing something wrong in the data pipeline. (The shape of my data is 112x124 for the gameboard and it has 6 layers representing different things in the game. Also, my input arrays only consist of zeros and ones to show if some object is there or not.)
I hope someone can help me to find out if I am doing something wrong.
Thanks in advance!

#### Version Information

* Deeplearning4j version: 0.9.1
* Platform information (OS, etc): Windows 7, 64-bit (Core i5, 8GB RAM)
* CUDA version, if used: Currently CPU only
* NVIDIA driver version, if in use: Currently CPU only

#### Code

**Used variables**
```
// fix random seed for reproducibility
protected static final int seed = 7;
protected static final Random rng = new Random(CNN.seed);

// parameters
private static final int batchSize = 128;
protected static final int numClasses = MOVE.values().length - 1;
private static final int epochs = 10;
private static final float testSize = 0.15f;

// input shape
protected static final int inputX = 112;
protected static final int inputY = 124;
protected static final int inputZ = 6;

// model parameters
private static final String savePath = "CNN/dl4j.zip";
private static final File saveFile = new File(CNN.savePath);
```

**Code for the CNN itself**
```
// build the CNN configuration
MultiLayerConfiguration conf = new NeuralNetConfiguration.Builder()
		.seed(CNN.seed)
		.iterations(1)
		.optimizationAlgo(OptimizationAlgorithm.STOCHASTIC_GRADIENT_DESCENT)
		.weightInit(WeightInit.RELU)
		.list()

		.layer(0, new ConvolutionLayer.Builder(5, 5)
				.padding(4, 4)
				.stride(1, 1)
				.nOut(32)
				.activation(Activation.RELU)
				.build())
		.layer(1, new SubsamplingLayer.Builder(PoolingType.MAX)
				.kernelSize(2, 2)
				.stride(2, 2)
				.build())

		.layer(2, new ConvolutionLayer.Builder(5, 5)
				.padding(4, 4)
				.stride(1, 1)
				.nOut(64)
				.activation(Activation.RELU)
				.build())
		.layer(3, new SubsamplingLayer.Builder(PoolingType.MAX)
				.kernelSize(2, 2)
				.stride(2, 2)
				.build())
				
		.layer(8, new DenseLayer.Builder()
				.activation(Activation.RELU)
				.nOut(1024)
				.build())
		.layer(9, new OutputLayer.Builder(LossFunction.MCXENT)
				.nOut(CNN.num_classes)
				.activation(Activation.SOFTMAX)
				.build())
		
		.setInputType(InputType.convolutionalFlat(CNN.inputY, CNN.inputX, CNN.inputZ))
		.backprop(true)
		.pretrain(false)
		.build();
		
// initialize CNN
MultiLayerNetwork model = new MultiLayerNetwork(conf);
model.init();
```

**Code to get the data in the right format and train the network**
```
// get data and reshape
float[][] xRaw = exporter.CNNXExport();
float[][] yRaw = exporter.CNNYExport();

int samples = xRaw.length; 

// split into train and test
INDArray features = Nd4j.create(xRaw);
INDArray labels = Nd4j.create(yRaw);
DataSet data = new DataSet(features, labels);
SplitTestAndTrain trainAndTest = data.splitTestAndTrain((int) (samples * (1 - CNN.testSize)), CNN.rng);

DataSet trainingData = trainAndTest.getTrain();
DataSet testData = trainAndTest.getTest();
		
// printing
System.out.println(samples + " Samples");
System.out.println("X Training Shape: " + Arrays.toString(trainingData.getFeatureMatrix().shape()));
System.out.println(trainingData.numExamples() + " Training Samples");
System.out.println(testData.numExamples() + " Test Samples");
		
// training
for (int i = 0; i < CNN.epochs; i++) {
	List<DataSet> batches = trainingData.batchBy(CNN.batchSize);
	for (int j = 0; j < batches.size(); j++) {
		this.model.fit(batches.get(j));
		System.out.println("Epoch " + (i + 1) + ": Batch " + (j + 1) + "/" + batches.size());
	}
	System.out.println("Completed Epoch " + (i + 1));
}

// evaluate the model on the test set
Evaluation eval = new Evaluation(CNN.numClasses);

List<DataSet> testInstances = testData.batchBy(1);
		
long timeTaken = 0;
long time = System.currentTimeMillis();
		
for (int i = 0; i < testInstances.size(); i++) {
	INDArray output = model.output(testInstances.get(i).getFeatures());
	
	timeTaken += (System.currentTimeMillis() - time);
	
	eval.eval(testInstances.get(i).getLabels(), output);
	
	time = System.currentTimeMillis();
}
		
System.out.println("\nAverage Output Time: " + (timeTaken / testInstances.size()));
		
System.out.println(eval.stats());

// save model to disk
try {
	ModelSerializer.writeModel(this.model, CNN.saveFile, true);
} catch (IOException e) {
	// TODO Auto-generated catch block
	e.printStackTrace();
}
```

**Comment by saudet** [comment]
It looks like you're using only 1 "image" as your DataSet? This isn't going to work...

**Comment by Raighar** [comment]
Why do you think so? I think I am splitting my data into different batches and iterate over all of them for training

**Comment by saudet** [comment]
`xRaw` is 2D, so that appears to be only one image, right?
